src/audio/mixin_mixout.c

```diff
@@ -248,235 +263,370 @@ struct mixout_source_info *find_mixout_source_info(struct mixed_data_info __spar
 	return NULL;
 }
 
-static void audio_stream_bytes_copy(struct audio_stream __sparse_cache *dst_stream, void *dst,
-				    const struct audio_stream __sparse_cache *src_stream,
-				    void *src, uint32_t size)
-{
-	int n;
-	uint8_t *pdst = (uint8_t *)dst;
-	uint8_t *psrc = (uint8_t *)src;
-
-	while (size) {
-		n = MIN(audio_stream_bytes_without_wrap(dst_stream, pdst), size);
-		n = MIN(audio_stream_bytes_without_wrap(src_stream, psrc), n);
-		memcpy_s(pdst, n, psrc, n);
-		size -= n;
-		pdst = audio_stream_wrap(dst_stream, pdst + n);
-		psrc = audio_stream_wrap(src_stream, psrc + n);
-	}
-}
-
 #if CONFIG_FORMAT_S16LE
-static void mix_s16(struct audio_stream __sparse_cache *sink, uint32_t start, uint32_t mixed_bytes,
-		    const struct audio_stream __sparse_cache *source, uint32_t size, uint16_t gain)
+/* Instead of using sink->channels and source->channels, sink_channel_count and
+ * source_channel_count are supplied as parameters. This is done to reuse the function to also
+ * mix entire stream. In this case the function is called with fake stream parameters: multichannel
+ * stream is represented as single channel and so the entire stream contents is mixed.
+ */
+static void mix_channel_s16(struct audio_stream __sparse_cache *sink, uint8_t sink_channel_index,
+			    uint8_t sink_channel_count, uint32_t start_frame, uint32_t mixed_frames,
+			    const struct audio_stream __sparse_cache *source,
+			    uint8_t source_channel_index, uint8_t source_channel_count,
+			    uint32_t frame_count, uint16_t gain)
 {
 	int16_t *dest, *src;
-	uint32_t bytes_to_mix, bytes_to_copy;
-	int i, n;
-
-	dest = audio_stream_wrap(sink, (uint8_t *)sink->w_ptr + start);
-	src = source->r_ptr;
-
-	assert(mixed_bytes >= start);
-	bytes_to_mix = MIN(mixed_bytes - start, size);
-	bytes_to_copy = size - bytes_to_mix;
-
-	while (bytes_to_mix != 0) {
-		n = MIN(audio_stream_bytes_without_wrap(sink, dest), bytes_to_mix);
-		n = MIN(audio_stream_bytes_without_wrap(source, src), n);
-		bytes_to_mix -= n;
-		n >>= 1;	/* divide by 2 */
-		if (gain == IPC4_MIXIN_UNITY_GAIN) {
-			for (i = 0; i < n; i++) {
+	uint32_t frames_to_mix, frames_to_copy;
+
+	/* audio_stream_wrap() is required and is done below in a loop */
+	dest = (int16_t *)sink->w_ptr + start_frame * sink_channel_count + sink_channel_index;
+	src = (int16_t *)source->r_ptr + source_channel_index;
+
+	assert(mixed_frames >= start_frame);
+	frames_to_mix = MIN(mixed_frames - start_frame, frame_count);
+	frames_to_copy = frame_count - frames_to_mix;
+
+	/* we do not want to use something like audio_stream_frames_without_wrap() here
+	 * as it uses stream->channels internally. Also we would like to avoid expensive
+	 * division operations.
+	 */
+	while (frames_to_mix) {
```

Comment:
The loop could be simplified by using something like audio_stream_get_avail_frames(), audio_stream_get_free_frames(). But these two use division internally which is about 40 times slower then just pointer incrementation. So I chose to track pointers not reaching beyond end of buffer instead of calculating frames and tracking processed frames.
No, infinite lope is not the case here.

Reply:
You're probably right, that the loop will never be infinite in practice, but it still looks risky to me. What it's doing there is something like
```
N=5;
while (N) {
	x=get_x();
	while (N && x) {
		N--;
	}
}
```
so, if `x` turns false before `N` is decremented to `0` the external loop will run forever. If we're absolutely sure, that the internal loop will be entered at least one for each external iteration, then we could at least replace the internal one with `do {} while (N && x)`

Reply:
The problem with replacing while() {...} with do {...} while() is that then do {...} while() version will look like it is possible to write to a buffer (to *dst) without a prior check if pointer is within buffer boundaries. Consider a buffer size == 0 (which should never happen, of course), in this case while() {} will turn into infinite loop and do {} while () will write beyond buffer boundary. Writing beyond buffer is much worse (harder to debug) then infinite loop. So do {...} while () version is not better then while () {...}.

Reply:
but you're saying that will never happen! :-) So, that's exactly my point. If there's any chance of an infinite loop there (which is the same as the problem you describe with my `do {} while ()` approach) - let's explicitly fix it now. If there's absolutely 0 such chance - then my `do {} while ()` isn't any worse and I'd say is better because it shows explicitly, that "we know what we're doing"

Reply:
(1) Zero buffer length should not happen, as if there is no space in sink buffer mix_channel_sXX() will not be called at all.

(2) For zero buffer length case (which should never happen), while() {} will result in infinite loop, but do {} while() will result in writing to memory beyond buffer region. Both are problems, but do {} while() is worse (harder to notice, could result in strange "magic" bugs). So it is none benefits in replacing while () {} with do {} while().
I understand, looking into nested while() {} loops pair triggers a suspicious is there an infinite loop possible. However, looking into do {} while() with memory write inside triggers the suspicion has the pointer been checked to be within buffer range. 

(3) The function x=get_x() from your example above is x = audio_stream_wrap(..., x) and the job of that function is to make sure x condition in while() on next line is true for first iteration.